File: 010/prob_10.clj
(ns prob-10
  (:use prob-03)  
)

(comment
  "Problem 10
08 February 2002

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million."
)

(defn sum-of-primes [n]
  (let [prim (primes n)]
    (reduce + prim)
  )
)

;;(prob-10/sum-of-primes 10)
;;17

;; (prob-10/sum-of-primes 2000000)
;;142913828922

;;(time (prob-10/sum-of-primes 2000000))
;;"Elapsed time: 14996.382256 msecs"
;;142913828922

File: java/src/java/Prob10.java
package java;
public class Prob10 {

	public static void main(String[] args) {
		System.out.println("The sum of all the primes below two million is "+ mainRoutine(2000000));
	}

	public static long mainRoutine(int threshold) {
		long sumOfPrime = 0;
		long counter = 0;
		while (counter < threshold) {
			if (isPrime(counter)) {
				sumOfPrime = sumOfPrime + counter;
			}
			counter++;
		}
		return sumOfPrime;
	}

	public static boolean isPrime(long n) {
		if (n <= 1) {
			return false;
		}
		if (n == 2) {
			return true;
		}
		for (int i = 2; i <= Math.sqrt(n) + 1; i++) {
			if (n % i == 0) {
				return false;
			}
		}
		return true;
	}
}

